File: src/Rglpk_solve.c
/* This is the GLPK C Interface
 */

#include "Rglpk.h"
#include <stdio.h>

// this is the solve function called from R
void R_glp_solve (int *lp_direction, int *lp_number_of_constraints,
		  int *lp_direction_of_constraints, double *lp_right_hand_side,
		  int *lp_number_of_objective_vars,
		  double *lp_objective_coefficients,
		  int *lp_objective_var_is_integer, 
		  int *lp_objective_var_is_binary,
		  int *lp_is_integer,                     //should be boolean
		  int *lp_number_of_values_in_constraint_matrix,
		  int *lp_constraint_matrix_i, int *lp_constraint_matrix_j,
		  double *lp_constraint_matrix_values,
		  int *lp_bounds_type, double *lp_bounds_lower,
		  double *lp_bounds_upper,
		  double *lp_optimum,
		  double *lp_objective_vars_values,
		  int *lp_verbosity, int *lp_status) {

  glp_prob *lp;
  int i, kl, ku;
  // create problem object 
  lp = glp_create_prob();

  // Turn on/off Terminal Output
  if(*lp_verbosity==1)
    glp_term_out(GLP_ON);
  else
    glp_term_out(GLP_OFF);
  
  // direction of optimization
  if(*lp_direction==1)
    glp_set_obj_dir(lp, GLP_MAX);
  else
    glp_set_obj_dir(lp, GLP_MIN);
  
  // is it a mixed integer problem? -- seems to be an R glpk function
    //if(lp_integer)
      //lpx_set_class(lp, LPX_MIP);
  // add rows to the problem object
  glp_add_rows(lp, *lp_number_of_constraints);
  for(i = 0; i < *lp_number_of_constraints; i++)
    switch(lp_direction_of_constraints[i]){
    case 1: 
      glp_set_row_bnds(lp, i+1, GLP_UP, 0.0, lp_right_hand_side[i]);
      break;
    case 2: 
      glp_set_row_bnds(lp, i+1, GLP_UP, 0.0, lp_right_hand_side[i]);
      break;
    case 3: 
      glp_set_row_bnds(lp, i+1, GLP_LO, lp_right_hand_side[i], 0.0);
      break;
    case 4: 
      glp_set_row_bnds(lp, i+1, GLP_LO, lp_right_hand_side[i], 0.0);
      break;
    case 5: 
      glp_set_row_bnds(lp, i+1, GLP_FX, lp_right_hand_side[i],
		       lp_right_hand_side[i]);
      break;
    }
  
  // add columns to the problem object
  glp_add_cols(lp, *lp_number_of_objective_vars);
  kl = ku = 0;
  for(i = 0; i < *lp_number_of_objective_vars; i++) {
    glp_set_col_bnds(lp, i+1, lp_bounds_type[i], lp_bounds_lower[i], lp_bounds_upper[i]);
    // set objective coefficients and integer if necessary
    glp_set_obj_coef(lp, i+1, lp_objective_coefficients[i]);
    if (lp_objective_var_is_integer[i])
      glp_set_col_kind(lp, i+1, GLP_IV);
    if (lp_objective_var_is_binary[i])
      glp_set_col_kind(lp, i+1, GLP_BV);
  }
  // load the matrix
  // IMPORTANT: as glp_load_matrix requires triplets as vectors of the
  // form: ia[1] ... ia[n], we have to pass the pointer to the adress
  // [-1] of the corresponding vector 

  glp_load_matrix(lp, *lp_number_of_values_in_constraint_matrix,
		  &lp_constraint_matrix_i[-1],
                  &lp_constraint_matrix_j[-1], &lp_constraint_matrix_values[-1]);

  // run simplex method to solve linear problem
  glp_simplex(lp, NULL);
  
  // retrieve status of optimization
  *lp_status = glp_get_status(lp);
  // retrieve optimum
  *lp_optimum = glp_get_obj_val(lp);
  // retrieve values of objective vars
  for(i = 0; i < *lp_number_of_objective_vars; i++) {
    lp_objective_vars_values[i] = glp_get_col_prim(lp, i+1);
  }
  if(*lp_is_integer) {
    glp_intopt(lp, NULL);
    // retrieve status of optimization
    *lp_status = glp_mip_status(lp);
    
    // retrieve MIP optimum
    *lp_optimum = glp_mip_obj_val(lp);
    // retrieve MIP values of objective vars
    for(i = 0; i < *lp_number_of_objective_vars; i++){
      lp_objective_vars_values[i] = glp_mip_col_val(lp, i+1);
    }
  }
  // delete problem object
  glp_delete_prob(lp);
}

File: R/Rglpk_solve.R
## the R-ported GNU Linear Programming kit
## solve function --- C Interface

Rglpk_solve_LP <-
function(obj, mat, dir, rhs, bounds = NULL, types = NULL, max = FALSE,
          control = list(), mrhs_i = NULL, mrhs_val = NULL,
          ...)
{
  ## validate direction of optimization
  if(!identical( max, TRUE ) && !identical( max, FALSE ))
    stop("'Argument 'max' must be either TRUE or FALSE.")
  direction_of_optimization <- as.integer(max)

  ## validate multi problem RHS
  if(! is.null(mrhs_i) ) {
    if (is.null(mrhs_val)) stop("If 'mrhs_i' is specified please also specify 'mrhs_val'.")
    if (!is.matrix(mrhs_val)) stop("Argument 'mrhs_val' has to be a matrix.")
    if (length(mrhs_i) != nrow(mrhs_val)) stop(sprintf("Length of mrhs_i (%d) has to equal nrows of mrhs_val (%d).", length(mrhs_i), nrow(mrhs_val)) )
    if (any(mrhs_i < 0 || mrhs_i >= length(rhs))) stop("Argument 'mrhs_i' must be a vector of integers between 0 and length(rhs)-1, inclusive.")
  }

  ## validate control list
  dots <- list(...)
  control[names(dots)] <- dots
  control <- .check_control_parameters( control )
  verb <- control$verbose

  ## match direction of constraints
  n_of_constraints <- length(dir)
  ## match relational operators to requested input
  direction_of_constraints <- match( dir, c("<", "<=", ">", ">=", "==") )

  if( any(is.na(direction_of_constraints)) )
    stop("Argument 'dir' must be either '<', '<=', '>', '>=' or '=='.")

  ## we need to verify that obj is a numeric vector
  ## FIXME: always use STMs?
  if(slam::is.simple_triplet_matrix(obj))
      obj <- as.matrix(obj)
  obj <- as.numeric(obj)
  n_of_objective_vars <- length( obj )

  constraint_matrix <- as.simple_triplet_matrix(mat)

  ## types of objective coefficients
  ## Default: "C"
  if(is.null(types))
    types <- "C"
  ## check if valid types
  if(any(is.na(match(types, c("I", "B", "C"), nomatch = NA))))
    stop("'types' must be either 'B', 'C' or 'I'.")
  ## replicate types to fit number of columns
  types <- rep(types, length.out = n_of_objective_vars)
  ## need a TRUE/FALSE integer/binary representation
  integers <- types == "I"
  binaries <- types == "B"

  ## do we have a mixed integer linear program?
  is_integer <- any( binaries | integers )

  ## bounds of objective coefficients
  bounds <- as.glp_bounds( as.list( bounds ), n_of_objective_vars )

  ## Sanity check: mat/dir/rhs
  if( !all(c(dim(mat)[ 1 ], length(rhs)) == n_of_constraints) )
      stop( "Arguments 'mat', 'dir', and/or 'rhs' not conformable." )
  ## Sanity check: mat, obj
  if( dim(mat)[ 2 ] != n_of_objective_vars )
      stop( "Arguments 'mat' and 'obj' not conformable." )

  ## call the C interface - this actually runs the solver
  x <- glp_call_interface(obj, n_of_objective_vars, constraint_matrix$i,
                          constraint_matrix$j, constraint_matrix$v,
                          length(constraint_matrix$v),
                          rhs, direction_of_constraints, n_of_constraints,
                          is_integer,
                          integers, binaries,
                          direction_of_optimization, bounds[, 1L],
                          bounds[, 2L], bounds[, 3L], verb,
                          1,         ## number of problems (at least 1)
                          c(), c(0), ## constraints
                          mrhs_i, mrhs_val) ## rhs

  solution <- x$lp_objective_vars_values
  ## are integer variables really integers? better round values
  solution[integers | binaries] <-
    round( solution[integers | binaries])
  ## match status of solution
  status <- as.integer(x$lp_status)
  if(control$canonicalize_status){
      ## 0 -> optimal solution (5 in GLPK) else 1
      status <- as.integer(status != 5L)
  }
  list(optimum = sum(solution * obj), solution = solution, status = status)
}

## this function calls the C interface
glp_call_interface <-
function(lp_objective_coefficients, lp_n_of_objective_vars,
         lp_constraint_matrix_i, lp_constraint_matrix_j, lp_constraint_matrix_v,
         lp_n_of_values_in_constraint_matrix, lp_right_hand_side,
         lp_direction_of_constraints, lp_n_of_constraints, lp_is_integer,
         lp_objective_var_is_integer, lp_objective_var_is_binary,
         lp_direction_of_optimization,
         lp_bounds_type, lp_bounds_lower, lp_bounds_upper,
         verbose,
         multi_number_of_problems, ## has to be at least 1
         multi_constraint_index, 
         multi_constraint_values,
         multi_rhs_index,
         multi_rhs_values ) 
{
  out <- .C("R_glp_solve",
            lp_direction_of_optimization= as.integer(lp_direction_of_optimization),
            lp_n_of_constraints         = as.integer(lp_n_of_constraints),
            lp_direction_of_constraints = as.integer(lp_direction_of_constraints),
            lp_right_hand_side          = as.double(lp_right_hand_side),
            lp_n_of_objective_vars      = as.integer(lp_n_of_objective_vars),
            lp_objective_coefficients   = as.double(lp_objective_coefficients),
            lp_objective_var_is_integer = as.integer(lp_objective_var_is_integer),
            lp_objective_var_is_binary  = as.integer(lp_objective_var_is_binary),
            lp_is_integer               = as.integer(lp_is_integer),
            lp_n_of_values_in_constraint_matrix = as.integer(lp_n_of_values_in_constraint_matrix),
            lp_constraint_matrix_i      = as.integer(lp_constraint_matrix_i),
            lp_constraint_matrix_j      = as.integer(lp_constraint_matrix_j),
            lp_constraint_matrix_values = as.double(lp_constraint_matrix_v),
            lp_bounds_type              = as.integer(lp_bounds_type),
            lp_bounds_lower             = as.double(lp_bounds_lower),
            ## lp_n_of_bounds_l            = as.integer(length(lp_lower_bounds_i)),
            lp_bounds_upper             = as.double(lp_bounds_upper),
            ## lp_n_of_bounds_u            = as.integer(length(lp_upper_bounds_i)),
            lp_optimum                  = double(multi_number_of_problems),
            lp_objective_vars_values    = double(lp_n_of_objective_vars * multi_number_of_problems),
            lp_verbosity                = as.integer(verbose),
            lp_status                   = integer(multi_number_of_problems),
            multi_number_of_problems    = as.integer(multi_number_of_problems),
            multi_number_of_constraint_values = 0, #length(multi_constraint_index),
            multi_constraint_index      = as.integer(multi_constraint_index),
            multi_constraint_values     = as.double(multi_constraint_values),
            multi_rhs_number_of_values  = 0, #length(multi_rhs_index),
            multi_rhs_index             = as.integer(multi_rhs_index),
            multi_rhs_values            = as.double(multi_rhs_values),
            NAOK = TRUE, PACKAGE = "rDEA")
  out
}

